tools/delete_imaging_upload.pl

```diff
@@ -0,0 +1,776 @@
+=cut
+sub deleteUploadsInDatabase {
+    my($dbh, $uploadsRef, $filesRef)= @_;
+    
+    $dbh->{'AutoCommit'} = 0;
+    
+    my $query = "DELETE FROM notification_spool WHERE ProcessID IN ("
+              . join(',', ('?') x keys %$uploadsRef)
+              . ")";
+    $dbh->do($query, undef, keys %$uploadsRef);
+    
+    my $tarchiveID = $filesRef->{'tarchive'}->[0]->{'TarchiveID'};
+    
+    $query = "DELETE FROM tarchive_files WHERE TarchiveID = ?";
+    $dbh->do($query, undef, $tarchiveID);
+    
+    $query = "DELETE FROM tarchive_series WHERE TarchiveID = ?";
+    $dbh->do($query, undef, $tarchiveID);
+    
+    if(@{ $filesRef->{'parameter_file'} }) {
+        my @parameterFileIDs = map { $_->{'FileID'} } @{ $filesRef->{'parameter_file'} };
+        $query  = sprintf(
+            "DELETE FROM parameter_file WHERE FileID IN (%s)",
+            join(',', ('?') x @parameterFileIDs)
+        );
+        $dbh->do($query, undef, @parameterFileIDs);
+    }
+
+    my @intermediaryFileIDs = map { $_->{'IntermedID'} } @{ $filesRef->{'files_intermediary'} };
+    if(@intermediaryFileIDs) {
+        $query = sprintf(
+            'DELETE FROM files_intermediary WHERE IntermedID IN (%s)', 
+            join(',', ('?') x @intermediaryFileIDs)
+        );
+        $dbh->do($query, undef, @intermediaryFileIDs);     
+
+        my @fileID = map { $_->{'FileID'} } @{ $filesRef->{'files_intermediary'} };
+        $query = sprintf(
+            "DELETE FROM files WHERE FileID IN (%s)",
+            join(',', ('?') x @fileID)
+        );
+        $dbh->do($query, undef, @fileID);
+    }
+            
+    $query = "DELETE FROM files WHERE TarchiveSource = ?";
+    $dbh->do($query, undef, $tarchiveID);
+    
+    $query = "DELETE FROM mri_protocol_violated_scans WHERE TarchiveID = ?";
+    $dbh->do($query, undef, $tarchiveID);
+    
+    $query = "DELETE FROM mri_violations_log WHERE TarchiveID = ?";
+    $dbh->do($query, undef, $tarchiveID);
+    
+    $query = "DELETE FROM MRICandidateErrors WHERE TarchiveID = ?";
+    $dbh->do($query, undef, $tarchiveID);
+    
+    $query = sprintf(
+        "DELETE FROM mri_upload WHERE UploadID IN (%s)",
+        join(',', map { '?' } keys %$uploadsRef)
+    );
+    $dbh->do($query, undef, keys %$uploadsRef);
+    
+    $query = "DELETE FROM tarchive WHERE TarchiveID = ?";
+    $dbh->do($query, undef, $tarchiveID);
+    
+    # If any of the upload to delete is the last upload that was part of the 
+    # session associated to it, then set the session's 'Scan_done' flag
+    # to 'N'.
+    my @sessionIDs = map { $uploadsRef->{'SessionID'} } keys %$uploadsRef;
+    $query = "UPDATE session s SET Scan_done = 'N'"
+           . " WHERE s.ID IN ("
+           . join(',', ('?') x @sessionIDs)
+           . ") AND (SELECT COUNT(*) FROM mri_upload m WHERE m.SessionID=s.ID) = 0";
+    $dbh->do($query, undef, map { $_->{'SessionID'} } @sessionIDs );
+    
+    $dbh->commit;
+    $dbh->{'AutoCommit'} = 0;
```

Comment:
@nicolasbrossard when I ran the script, I got 2 SQL failures:
```
DBD::mysql::db do failed: Cannot delete or update a parent row: a foreign key constraint fails (`cmadjar_preventad`.`parameter_file`, CONSTRAINT `FK_parameter_file_1` FOREIGN KEY (`FileID`) REFERENCES `files` (`FileID`)) at delete_imaging_upload.pl line 719.
DBD::mysql::db do failed: Cannot delete or update a parent row: a foreign key constraint fails (`cmadjar_preventad`.`files`, CONSTRAINT `FK_files_TarchiveID` FOREIGN KEY (`TarchiveSource`) REFERENCES `tarchive` (`TarchiveID`)) at delete_imaging_upload.pl line 737.
```

Therefore, I thought all the DB entries would have remained due to the failures (that I can't explain) but I no longer have data in the tarchive_series, tarchive_files, mri_upload tables so I guess the safe delete in the tables is not working?

Reply:
also, why is $dbh->{'AutoCommit'} = 0; set at the beginning and at the end of the function? (not sure how the auto commit works)

Reply:
Ok for the AutoCommit thing. It should be reset to 1 at the end of the method.